File: backend/app/services/vector_service.py
import math
import json
from typing import List, Dict, Any, Optional
from sqlalchemy.orm import Session
from app.models import Chunk, Document

def cosine_similarity(v1: List[float], v2: List[float]) -> float:
    """
    Computes cosine similarity between two float vectors.
    """
    dot_product = sum(a * b for a, b in zip(v1, v2))
    norm_a = sum(a * a for a in v1)
    norm_b = sum(b * b for b in v2)
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return dot_product / (math.sqrt(norm_a) * math.sqrt(norm_b))

def search_similar_chunks(
    db: Session,
    user_id: int,
    query_vector: List[float],
    limit: int = 5,
    document_id: Optional[int] = None
) -> List[Dict[str, Any]]:
    """
    Searches user's indexed chunks in SQLite and returns top matches sorted by similarity.
    """
    # 1. Fetch chunks belonging to the user
    query = db.query(Chunk).filter(Chunk.user_id == user_id)
    if document_id:
        query = query.filter(Chunk.document_id == document_id)
        
    chunks = query.all()
    if not chunks:
        return []

    # Map document IDs to names to construct source metadata
    doc_ids = list(set(c.document_id for c in chunks))
    documents = db.query(Document).filter(Document.id.in_(doc_ids)).all()
    doc_map = {doc.id: doc.name for doc in documents}

    # 2. Score similarity
    scored_results = []
    for chunk in chunks:
        try:
            # Deserialize the stored embedding vector JSON string
            chunk_vector = json.loads(chunk.embedding)
            score = cosine_similarity(query_vector, chunk_vector)
            
            scored_results.append({
                "chunk_id": chunk.id,
                "content": chunk.content,
                "page_number": chunk.page_number,
                "document_id": chunk.document_id,
                "document_name": doc_map.get(chunk.document_id, "Unknown Document"),
                "score": score
            })
        except Exception as error:
            print(f"Failed to score chunk {chunk.id}: {error}")
            continue

    # 3. Sort by score descending and return top limit
    scored_results.sort(key=lambda x: x["score"], reverse=True)
    return scored_results[:limit]

File: backend/src/services/vectorService.ts
import Chunk from '../models/Chunk';
import Document from '../models/Document';

export interface SimilarityResult {
  chunkId: string;
  content: string;
  pageNumber: number;
  documentId: string;
  documentName: string;
  score: number;
}

/**
 * Calculates the cosine similarity between two vectors
 */
export const cosineSimilarity = (vecA: number[], vecB: number[]): number => {
  let dotProduct = 0;
  let normA = 0;
  let normB = 0;
  const length = Math.min(vecA.length, vecB.length);
  
  for (let i = 0; i < length; i++) {
    dotProduct += vecA[i] * vecB[i];
    normA += vecA[i] * vecA[i];
    normB += vecB[i] * vecB[i];
  }
  
  if (normA === 0 || normB === 0) return 0;
  return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
};

/**
 * Performs a vector search over the user's document chunks in MongoDB using Cosine Similarity
 */
export const searchSimilarChunks = async (
  userId: string,
  queryVector: number[],
  limit: number = 5,
  documentId?: string
): Promise<SimilarityResult[]> => {
  // 1. Build the query filter
  const filter: any = { userId };
  if (documentId) {
    filter.documentId = documentId;
  }

  // 2. Fetch the chunks and populate their document names
  const chunks = await Chunk.find(filter).lean();
  if (chunks.length === 0) return [];

  // Fetch document details to map document names
  const docIds = Array.from(new Set(chunks.map(c => c.documentId.toString())));
  const docs = await Document.find({ _id: { $in: docIds } }).lean();
  const docMap = new Map(docs.map(d => [d._id.toString(), d.name]));

  // 3. Compute cosine similarity for each chunk
  const scoredChunks: SimilarityResult[] = chunks.map((chunk: any) => {
    const score = cosineSimilarity(queryVector, chunk.embedding);
    return {
      chunkId: chunk._id.toString(),
      content: chunk.content,
      pageNumber: chunk.pageNumber,
      documentId: chunk.documentId.toString(),
      documentName: docMap.get(chunk.documentId.toString()) || 'Unknown Document',
      score,
    };
  });

  // 4. Sort by score descending and take the top k
  scoredChunks.sort((a, b) => b.score - a.score);
  return scoredChunks.slice(0, limit);
};
